### PR TITLE
Token offsets for transformer tokenizers

### PR DESCRIPTION
**Pull request by dirkgr**
I had myself convinced that having these offsets would be helpful when building RC models.  I'm no longer sure that's the case, but this might still be a net-positive addition to the library.

This code is a bit hacky. It would be more satisfying to do this is with a real sequence alignment algorithm. But I'm not sure the complexity and runtime cost of doing that is really worth it.

### CODE REVIEW

**Comment by dirkgr** [comment]
@brendan-ai2, this slows tokenization by a factor of 1.6 or so. How performance conscious are we? Is that fine?

**Comment by dirkgr** [comment]
@eladsegal's approach is no doubt more thorough, and we could port it into here, but if Huggingface is going to get this capability natively anyways, and while it's not clear I really need this for better RC, I'd rather merge this better-than-nothing solution and move on.

**Comment by schmmd** [comment]
It seems like we should merge this, leave it off by default as it adds some overhead, and perhaps change that behavior at a later point.

File: allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py [hunk 1]
```diff
@@ -92,3 +96,137 @@ def test_splits_uncased_bert(self):
...
+        idxs = [t.idx for t in tokenized]
+        assert idxs == expected_idxs
+
+    def test_token_idx_roberta(self):
+        sentence = "A, naïve <mask> AllenNLP sentence."
+        expected_tokens = [
+            "<s>",
+            "A",
+            ",",
+            "ĠnaÃ¯ve",  # What happened here?
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py]: We know what happened there. `transformers` mangled this one badly. I'll change the comment to reflect that.

File: allennlp/data/tokenizers/pretrained_transformer_tokenizer.py [hunk 2]
```diff
@@ -59,19 +59,26 @@ def __init__(
...
         self._stride = stride
         self._truncation_strategy = truncation_strategy
+        self._calculate_character_offsets = calculate_character_offsets
+
+    def _tokenizer_lowercases(self):
+        if hasattr(self._tokenizer, "do_lower_case"):
+            return self._tokenizer.do_lower_case
+        if hasattr(self._tokenizer, "basic_tokenizer"):
+            return self._tokenizer.basic_tokenizer.do_lower_case
+        return False
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp/data/tokenizers/pretrained_transformer_tokenizer.py]: I replaced this with some detection code that runs in the constructor. I think that's safer than checking a list of previously known types, and safer than what I had before. At least this new version has a chance of getting it right when they add a new tokenizer.

**Pull request merged**